nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,570 @@
+  carrot_msg->header = carrot_pose.header;
+  carrot_msg->point.x = carrot_pose.pose.position.x;
+  carrot_msg->point.y = carrot_pose.pose.position.y;
+  carrot_msg->point.z = 0.01;  // publish right over map to stand out
+  return carrot_msg;
+}
+
+geometry_msgs::msg::TwistStamped RegulatedPurePursuitController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & speed)
+{
+  // Transform path to robot base frame
+  auto transformed_plan = transformGlobalPlan(pose);
+
+  // Find look ahead distance and point on path and publish
+  const double lookahead_dist = getLookAheadDistance(speed);
+  auto carrot_pose = getLookAheadPoint(lookahead_dist, transformed_plan);
+  carrot_pub_->publish(std::move(createCarrotMsg(carrot_pose)));
+
+  double linear_vel, angular_vel;
+
+  // Find distance^2 to look ahead point (carrot) in robot base frame
+  // This is the chord length of the circle
+  const double carrot_dist2 =
+    (carrot_pose.pose.position.x * carrot_pose.pose.position.x) +
+    (carrot_pose.pose.position.y * carrot_pose.pose.position.y);
+
+  // Find curvature of circle (k = 1 / R)
+  double curvature = 0.0;
+  if (carrot_dist2 > 0.001) {
+    curvature = 2.0 * carrot_pose.pose.position.y / carrot_dist2;
+  }
+
+  // Apply curvature to angular velocity
+  linear_vel = desired_linear_vel_;
+  angular_vel = desired_linear_vel_ * curvature;
+
+  // Make sure we're in compliance with basic constraints
+  double angle_to_path;
+  if (shouldRotateToPath(carrot_pose, angle_to_path)) {
+    rotateToHeading(linear_vel, angular_vel, angle_to_path, speed);
+  } else {
+    applyConstraints(
+      linear_vel, fabs(lookahead_dist - sqrt(carrot_dist2)),
+      lookahead_dist, curvature, speed,
+      costAtPose(pose.pose.position.x, pose.pose.position.y));
+  }
+
+  // Collision checking on this velocity heading
+  if (isCollisionImminent(pose, carrot_pose, curvature, linear_vel, angular_vel)) {
+    RCLCPP_ERROR(logger_, "Collision imminent!");
+    throw std::runtime_error("RegulatedPurePursuitController detected collision ahead!");
+  }
+
+  // populate and return message
+  geometry_msgs::msg::TwistStamped cmd_vel;
+  cmd_vel.header = pose.header;
+  cmd_vel.twist.linear.x = linear_vel;
+  cmd_vel.twist.angular.z = angular_vel;
+  return cmd_vel;
+}
+
+bool RegulatedPurePursuitController::shouldRotateToPath(
+  const geometry_msgs::msg::PoseStamped & carrot_pose, double & angle_to_path)
+{
+  // Whether we should rotate robot to rough path heading
+  angle_to_path = atan2(carrot_pose.pose.position.y, carrot_pose.pose.position.x);
+  return use_rotate_to_heading_ && fabs(angle_to_path) > rotate_to_heading_min_angle_;
+}
+
+void RegulatedPurePursuitController::rotateToHeading(
+  double & linear_vel, double & angular_vel,
+  const double & angle_to_path, const geometry_msgs::msg::Twist & curr_speed)
+{
+  // Rotate in place using max angular velocity / acceleration possible
+  linear_vel = 0.0;
```

Comment:
Just a question. Linear velocity in one moment changes from some value (may be big one) to zero. Robot moment-ally stops. This might cause a robot drift. Do we consider in the model of Controller these possible linear drifts that will cause undesirable frame shifts?

Reply:
I don't understand what you mean by robot drift

Reply:
This is also for rotating to heading if the heading is off by a ton, this isn't called in normal operations. Only on starting of a new path when there could be a huge diff needing correction

Reply:
Maybe the "drifting" term isn't correct. I mean robot slippage when it stops with braking distances trying to decrease speed value to 0 in one time. Something like here on this YouTube video: https://youtu.be/TdwuftIUwYY?t=131. Indeed, this is not a car robot (although the mobile robot can move quickly too), but the drifting/slipping principle I want to point is good shown here.

Reply:
Ah, well robots have kinematically feasible linear velocity transitions in this controller so that isn't an issue. Also its expected that robots make sure commands are valid before executing, its on all robots that I'm aware of.